### ISSUE TITLE
Ensure that macro code doesn't leak into released applications

### ISSUE THREAD

**Issue by jakemac53** [dart-lang/language#1892]
With the macro proposal as it stands today application code ends up with transitive imports to the macro libraries and all their supporting code - this is because the macro classes themselves are what those users use to apply macros.

Compare that to a solution like `build_runner` where the builder code is decoupled from the annotations that cause it to be applied - this means application code doesn't end up with `analyzer` or `build` dependencies, etc.

One approach to this problem is to just rely on tree shaking, but I don't believe that is a sufficient solution. At a minimum it will involve significant additional code bloat for debug builds, which will degrade the developer experience.

https://github.com/dart-lang/language/pull/1831 was one attempt to solve this issue that we decided against, due to complexity as well as a loss of testability.

Possibly we should decouple the macro implementation from the annotation you use to apply it, similar to how build_runner works?

**Comment by jakemac53** [comment]
> If it _does_ turn out to be a problem, I still think that "compile time" should be a property of the _import_ and not the _library_. In other words, I like the idea (that @kevmoo initially suggested in #1360) of something like:
> 
> ```
> const import 'some_macro_lib.dart';
> ```

That was my first intuition for how to deal with this, and I think we could make that approach work. But I liked controlling this at the  library level instead for a few reasons:

## We don't have to introduce a new kind of import for people to learn

This isn't a huge deal, I don't think it adds a ton of complexity for users.

## You can't accidentally use a normal import and bloat the app

I think we could mostly resolve by _only_ allowing macro applications to invoke constructors from const imports.

## Macros and other code could be safely exposed through a single import.

I do think this was a really nice part about that approach. People will end up needing to have 2 imports for any library that has both runtime code and macros.

---

But it does come with the downside of making testing more difficult. I think for users of macros the library approach is better, and for authors of macros the import approach is better.

**Comment by lrhn** [comment]
> Can you clarify what you mean by this?
> 
> I my macro library proposal I allowed you to create a normal library which exports both macro libraries and normal libraries, not sure if this is related.

Macros code is plain Dart code, so you'd likely be using a lot of helper libraries and support classes. Not all of that is uniquely macro code. So, like most other packages, you're likely to have dozens of library files in `src/` of your package, and maybe export a bunch of them through a top-level library.

You say that a normal library can export a macro library (without becoming one). That worries me in two ways. 

First, you don't export libraries, you export declarations, so if the designation of being a "macro library" is really on the library, then it's not clear what exporting declarations from that library does with the "being a macro library" part.

Also, I was assuming that the top-level library which is exporting the macro definitions would need to be a macro library itself. If importing a macro library is special, in that use them as macro annotations in otherwise incomplete libraries, then a "normal" library containing a macro definition seems like it shouldn't work. (That's why I assumed the top-level library would be the macro library, and it would just be exporting normal libraries as macros).

Is the underlying idea here to have *multi-stage imports*? Where you have one import, which imports both macro declarations to be used in "stage one" (running macros), and further declarations to be used in "stage two" (compiling after running the macros and removing the stage-one code)?
That does make sense, and with some kind of marker on the export, it might just work.

Consider a "macro library" like:
```dart
export "src/macro_decl.dart";  
late export "src/helper.dart";
```
which you can then `const import "package:macro/macro.dart";` and get the macro available, and also immediately add the helper declarations to library being generated. (And each *declaration* exported would then remember whether it's available now or later).

So, basically, what does it really mean to be "a macro library"?

I worry that making the distinction between normal and macro libraries too big is not worth the complexity it adds. I'd prefer if being used as a macro is a *role*, not necessarily an *identity*.

**Comment by munificent** [comment]
> Macros are like a separate _program_ - they need to be runnable by themselves, which does mean that their code must be complete and valid Dart at the time they are run.

This is true, but all that's needed to ensure that is to prohibit import cycles. In the absence of a cycle, every Dart library can be completely separately compiled.

**Comment by davidmorgan** [comment]
This should be resolved with #3728
